source/write/transactions.txt

```diff
@@ -0,0 +1,194 @@
+When run, the code performs the following actions: 
+
+1. Starts a session from the client using the ``start_session()`` method
+#. Defines a callback function that specifies the operations to perform during the transaction
+#. Creates an option object to set the write preference for the transaction operations
+#. Calls the ``with_transactions()`` method to manage the transaction, passing the callback and option object as arguments
+
+.. literalinclude:: /includes/write/transactions.cpp
+   :language: cpp
+   :dedent:
+   :start-after: start-callback-api
+   :end-before: end-callback-api
+
+.. _cpp-core-api: 
+
+Core API 
+~~~~~~~~
+
+Use the core API for more manual control of the transaction lifecycle. The core API doesn't incorporate error handling logic for ``TransientTransactionError`` and ``UnknownTransactionCommitResult`` errors,
+and instead provides the flexibility to incorporate custom error handling. To learn more about these errors, 
+see the :manual:`Transaction Error Handling </core/transactions-in-applications/#transaction-error-handling/>` section in the {+mdb-server+} manual.
+
+The core API requires explicit calls to start, commit, and abort the transaction using methods provided by the ``mongocxx::client_session`` interface.
+The following table describes these methods:  
+
+.. list-table::
+   :widths: 25 75
+   :header-rows: 1
+
+   * - Method
+     - Description
+
+   * - ``start_transaction()``
+     - | Starts a new transaction on the current client session. Accepts an optional ``mongocxx::options::transaction`` 
+         instance as an argument to set options. For a full list of options, see `mongocxx::options::transaction <{+api+}/classmongocxx_1_1options_1_1transaction.html>`__
+         in the API documentation.
+       |
+       | Raises an exception if the options are misconfigured, if there are network or other transient failures, or if there 
+         are other errors such as a session with a transaction already in progress. If an error is returned with the ``TransientTransactionError`` label, 
+         you can end the transaction and then retry it with the expectation that it will succeed. 
+       |
+       | To learn more about this method, see the :manual:`startTransaction()
+         </reference/method/Session.startTransaction/>` guide in the {+mdb-server+} manual.
+
+   * - ``commit_transaction()``
+     - | Commits the active transaction on the current client session.
+       |
+       | Raises an exception if options are misconfigured, if there are network or other transient failures,
+         or if there are other errors such as a session with no transaction in progress. If an error is returned with the ``UnknownTransactionCommitResult`` label,
+         you can end the transaction and then retry it with the expectation that it will succeed when the committed transaction satisfies the set write concern. 
```

Comment:
Q: what does "committed transaction [satisfying] the set write concern" look like? Can't quite wrap my head around it

Reply:
I adapted this line from the [Rust Transactions page](https://www.mongodb.com/docs/drivers/rust/v2.8/fundamentals/transactions/). I think the write concern states the number of replica set members that the transaction operations should write to when it performs a write operation (e.g. write to the "majority" of replica set members). So the UnknownTransactionCommitResult error indicates that the driver doesn't know whether or not the write operations have been replicated across the preferred number of replica set members. 

So the logic here is that the active transaction will be re-committed until it can be certain that it has been replicated across enough sets to "satisfy" the set write concern (e.g. replicated across the majority).